ReplayKitExample/BroadcastExtension/ExampleCoreAudioDevice.m

```diff
@@ -0,0 +1,627 @@
+                                                      UInt32 busNumber,
+                                                      UInt32 numFrames,
+                                                      AudioBufferList *bufferList) {
+    assert(bufferList->mNumberBuffers == 1);
+    assert(bufferList->mBuffers[0].mNumberChannels <= 2);
+    assert(bufferList->mBuffers[0].mNumberChannels > 0);
+
+    ExampleCoreAudioContext *context = (ExampleCoreAudioContext *)refCon;
+    int8_t *audioBuffer = (int8_t *)bufferList->mBuffers[0].mData;
+    UInt32 audioBufferSizeInBytes = bufferList->mBuffers[0].mDataByteSize;
+
+    // Render silence if there are temporary mismatches between CoreAudio and our rendering format.
+    if (numFrames > context->maxFramesPerBuffer) {
+        NSLog(@"Can handle a max of %u frames but got %u.", (unsigned int)context->maxFramesPerBuffer, (unsigned int)numFrames);
+        *actionFlags |= kAudioUnitRenderAction_OutputIsSilence;
+        memset(audioBuffer, 0, audioBufferSizeInBytes);
+        return noErr;
+    }
+
+    // Pull decoded, mixed audio data from the media engine into the AudioUnit's AudioBufferList.
+    assert(numFrames <= context->maxFramesPerBuffer);
+    assert(audioBufferSizeInBytes == (bufferList->mBuffers[0].mNumberChannels * kAudioSampleSize * numFrames));
+    TVIAudioDeviceReadRenderData(context->deviceContext, audioBuffer, audioBufferSizeInBytes);
+    return noErr;
+}
+
+OSStatus ExampleCoreAudioDeviceRecordCallback(CMSampleBufferRef sampleBuffer) {
+    if (!capturingContext || !capturingContext->deviceContext) {
+        return noErr;
+    }
+
+    CMBlockBufferRef blockBuffer = CMSampleBufferGetDataBuffer(sampleBuffer);
+
+    AudioBufferList bufferList;
+    CMSampleBufferGetAudioBufferListWithRetainedBlockBuffer(sampleBuffer,
+                                                            NULL,
+                                                            &bufferList,
+                                                            sizeof(bufferList),
+                                                            NULL,
+                                                            NULL,
+                                                            kCMSampleBufferFlag_AudioBufferList_Assure16ByteAlignment,
+                                                            &blockBuffer);
+
+    int8_t *audioBuffer = (int8_t *)bufferList.mBuffers[0].mData;
+    UInt32 audioBufferSizeInBytes = bufferList.mBuffers[0].mDataByteSize;
+
+    bool nonNullFound = false;
+    for (int i = 0; i< audioBufferSizeInBytes; i++) {
+        if (*(audioBuffer+i) != 0) {
+            nonNullFound = true;
+            break;
+        }
+    }
+    if (!nonNullFound) {
+        return noErr;
+    }
+
+    TVIAudioDeviceWriteCaptureData(capturingContext->deviceContext, (int8_t *)audioBuffer, audioBufferSizeInBytes);
+
+    CFRelease(blockBuffer);
+    return noErr;
+}
```

Comment:
Revisit this code, without the assumption of two audio sources. We may be able to eliminate the silence checks.